pennylane/pauli/conversion.py

```diff
@@ -0,0 +1,110 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Utility functions to convert between ``~.PauliSentence`` and other PennyLane operators.
+"""
+from operator import matmul
+from itertools import product
+from typing import Union
+from functools import reduce
+
+import numpy as np
+
+from pennylane.ops import Hamiltonian
+from .pauli_arithmetic import PauliWord, PauliSentence, I, X, Y, Z, mat_map, op_map
+
+
+def decompose(
+    H, hide_identity=False, wire_order=None, pauli=False
+) -> Union[Hamiltonian, PauliSentence]:
+    r"""Decomposes a Hermitian matrix into a linear combination of Pauli operators.
+
+    Args:
+        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
+        hide_identity (bool): does not include the Identity observable within
+            the tensor products of the decomposition if ``True``.
+        wire_order (list[Union[int, str]]): the ordered list of wires with respect
+            to which the operator is represented as a matrix.
+        pauli (bool): return a PauliSentence instance if ``True``.
+
+    Returns:
+        Union[~.Hamiltonian, ~.PauliSentence]: the matrix decomposed as a linear combination
+            of Pauli operators, either as a ``~.Hamiltonian`` or ``~.PauliSentence`` instance.
+
+    **Example:**
+
+    We can use this function to compute the Pauli operator decomposition of an arbitrary Hermitian
+    matrix:
+
+    >>> A = np.array(
+    ... [[-2, -2+1j, -2, -2], [-2-1j,  0,  0, -1], [-2,  0, -2, -1], [-2, -1, -1,  0]])
+    >>> H = decompose(A)
+    >>> print(H)
+    (-1.0) [I0 I1]
+    + (-1.5) [X1]
+    + (-0.5) [Y1]
+    + (-1.0) [Z1]
+    + (-1.5) [X0]
+    + (-1.0) [X0 X1]
+    + (-0.5) [X0 Z1]
+    + (1.0) [Y0 Y1]
+    + (-0.5) [Z0 X1]
+    + (-0.5) [Z0 Y1]
+
+    We can return a ``~.PauliSentence`` instance by using the keyword argument ``pauli=True``:
+
+    >>> ps = decompose(A, pauli=True)
+    >>> print(ps)
+    """
+    n = int(np.log2(len(H)))
+    N = 2**n
+
+    if wire_order is None:
+        wire_order = range(n)
+
+    if H.shape != (N, N):
+        raise ValueError("The matrix should have shape (2**n, 2**n), for any qubit number n>=1")
+
+    if not np.allclose(H, H.conj().T):
+        raise ValueError("The matrix is not Hermitian")
+
+    obs_lst = []
+    coeffs = []
+
+    for term in product([I, X, Y, Z], repeat=n):
+        matrices = [mat_map[i] for i in term]
+        coeff = np.trace(reduce(np.kron, matrices) @ H) / N
+        coeff = np.real_if_close(coeff).item()
+
+        if not np.allclose(coeff, 0):
+            obs_term = (
+                [(o, w) for w, o in zip(wire_order, term) if o != I]
+                if hide_identity
+                else [(o, w) for w, o in zip(wire_order, term)]
+            )
+
+            if obs_term:
+                coeffs.append(coeff)
+                obs_lst.append(obs_term)
```

Comment:
Could we just directly create a `PauliWord` and add it to an existing `PauliSentence` at this point?

Reply:
Yes we could, is the idea then to construct the `PauliSentence` and the list of coeffs, terms and then depending on the flag, return one or Hamiltonian(other)? I don't have a preference for either option. I would guess that building it iteratively vs all at once should have the same total overhead in the end.